### PR TITLE
Fixed #28488 -- Allowed error handlers to access CSRF tokens.

### PR DESCRIPTION
**Pull request by apollo13**
This should fix ticket 28488. The main issue was that error handlers could access CSRF tokens and not run through `process_view`. The patch just masks the problem by ensuring that the handling already happens in `process_request`. All in all this means (for stable CSRF operations) that csrf middleware should come relatively early in the middleware chain -- but I think this is the best we can do.

Needs backport to at least 1.11

### CODE REVIEW

File: tests/csrf_tests/tests.py [hunk 1]
```diff
@@ -723,9 +749,22 @@ def test_https_reject_insecure_referer(self):
...
             'Referer checking failed - Referer is insecure while host is secure.',
             status_code=403,
         )
+
+
+@override_settings(ROOT_URLCONF='csrf_tests.urls', DEBUG=False)
+class CsrfInErrorHandlingViewsTests(SimpleTestCase):
+    def test_csrf_token_on_404_stays_constant(self):
+        response = self.client.get('/does not exist/')
+        # status code 999 is used to ensure that the error handler was called
```
Comment (claudep) [inline tests/csrf_tests/tests.py]: 999 or 599?

**Comment by claudep** [comment]
Looks good, could you add the release note for 1.11 in the patch?

File: tests/csrf_tests/tests.py [hunk 2]
```diff
@@ -583,11 +605,13 @@ def test_bare_secret_accepted_and_replaced(self):
         """
         The csrf token is reset from a bare secret.
         """
+        middleware = self.mw
```
Comment (timgraham) [inline tests/csrf_tests/tests.py]: Is there a reason this test uses a different style (local variable assignment)?

File: tests/csrf_tests/tests.py [hunk 3]
```diff
@@ -39,6 +39,7 @@ class CsrfViewMiddlewareTestMixin:
     """
 
     _csrf_id = _csrf_id_cookie = '1bcdefghij2bcdefghij3bcdefghij4bcdefghij5bcdefghij6bcdefghijABCD'
+    mw = CsrfViewMiddleware()
```
Comment (timgraham) [inline tests/csrf_tests/tests.py]: Is this refactoring a needed part of this patch? Can in be done in a separate commit that precedes the bug fix? I see that at least the `process_request()` additions have to be in the fix.

File: tests/csrf_tests/tests.py [hunk 4]
```diff
@@ -655,7 +679,9 @@ def test_no_session_on_request(self):
             'SessionMiddleware must appear before CsrfViewMiddleware in MIDDLEWARE.'
         )
         with self.assertRaisesMessage(ImproperlyConfigured, msg):
-            CsrfViewMiddleware().process_view(HttpRequest(), None, (), {})
+            req = HttpRequest()
```
Comment (timgraham) [inline tests/csrf_tests/tests.py]: Move the first two lines out of the context manager so it's more clear which line is raising the exception.

File: docs/releases/1.11.6.txt [hunk 5]
```diff
@@ -11,3 +11,6 @@ Bugfixes
 
 * Made the ``CharField`` form field convert whitespace-only values to the
   ``empty_value`` when ``strip`` is enabled (:ticket:`28555`).
+
+* Ensured that CSRF processing works as expected when an error handler
```
Comment (timgraham) [inline docs/releases/1.11.6.txt]: suggested wording: Fixed a regression in Django 1.9 where a custom view error handler such as ``handler404`` that accesses ``csrf_token`` (for example, by using ``django.shortcuts.render()``) could cause CSRF verification failures on other pages

File: tests/csrf_tests/views.py [hunk 6]
```diff
@@ -28,3 +29,8 @@ def non_token_view_using_request_processor(request):
     context = RequestContext(request, processors=[csrf])
     template = Template('')
     return HttpResponse(template.render(context))
+
+
+def error_handler(request, **args):
```
Comment (timgraham) [inline tests/csrf_tests/views.py]: I'd use a more specific name and/or docstring that describes what's special about this error handler to trigger the regression since I could imagine future tests may want to use error handlers that have different features. Similar for the `urls.py` and test naming and/or docstring.

Comment (timgraham) [inline tests/csrf_tests/views.py]: **kwargs?

File: django/middleware/csrf.py [hunk 7]
```diff
@@ -194,15 +194,18 @@ def _set_token(self, request, response):
             # Set the Vary header since content varies with the CSRF cookie.
             patch_vary_headers(response, ('Cookie',))
 
-    def process_view(self, request, callback, callback_args, callback_kwargs):
-        if getattr(request, 'csrf_processing_done', False):
-            return None
-
+    def process_request(self, request, *args, **kwargs):
```
Comment (timgraham) [inline django/middleware/csrf.py]: I think `*args, **kwargs` are unexpected.

File: django/middleware/csrf.py [hunk 8]
```diff
@@ -194,15 +194,18 @@ def _set_token(self, request, response):
...
         csrf_token = self._get_token(request)
         if csrf_token is not None:
             # Use same token next time.
             request.META['CSRF_COOKIE'] = csrf_token
 
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        if getattr(request, 'csrf_processing_done', False):
+            return None
+
+        csrf_token = request.META.get('CSRF_COOKIE')
```
Comment (timgraham) [inline django/middleware/csrf.py]: How about putting this right above where it's first used rather than far about it? (`if csrf_token is None:`)

File: docs/releases/1.11.6.txt (hunk 5 above)
Comment (apollo13) [reply docs/releases/1.11.6.txt]: Using `render` shouldn't actually cause this issue since the token is used lazily.

File: tests/csrf_tests/tests.py (hunk 3 above)
Comment (apollo13) [reply tests/csrf_tests/tests.py]: No, it is not necessarily needed, but the patch is small enough (and I am honestly to lazy to revert them and move them into an extra commit).
